controllers/state_of_the_world.go

```diff
@@ -0,0 +1,149 @@
+		controller.WithRunnable("ratelimitpolicy watcher", controller.Watch(&kuadrantv1beta2.RateLimitPolicy{}, kuadrantv1beta2.RateLimitPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("configmap watcher", controller.Watch(&corev1.ConfigMap{}, controller.ConfigMapsResource, kuadrantNamespace, controller.FilterResourcesByLabel[*corev1.ConfigMap]("kuadrant.io/topology=true"))),
+		controller.WithPolicyKinds(
+			kuadrantv1alpha1.DNSPolicyKind,
+			kuadrantv1alpha1.TLSPolicyKind,
+			kuadrantv1beta2.AuthPolicyKind,
+			kuadrantv1beta2.RateLimitPolicyKind,
+		),
+		controller.WithObjectKinds(
+			kuadrantv1beta1.KuadrantKind,
+			schema.GroupKind{Group: corev1.GroupName, Kind: "ConfigMap"}),
+		controller.WithObjectLinks(
+			kuadrantv1beta1.LinkKuadrantToGatewayClasses,
+			kuadrantv1beta1.LinkKuadrantToTopologyConfigMap,
+		),
+		controller.WithReconcile(buildReconciler(client)),
+	}
+
+	return controller.NewController(controllerOpts...)
+}
+
+func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+	reconciler := &controller.Workflow{
+		Precondition: NewEventLogger().Log,
+		Tasks: []controller.ReconcileFunc{
+			NewTopologyFileReconciler(client).Reconcile,
+		},
+	}
+
+	return reconciler.Run
+}
+
+type TopologyFileReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewTopologyFileReconciler(client *dynamic.DynamicClient) *TopologyFileReconciler {
+	return &TopologyFileReconciler{Client: client}
+}
+
+func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology) {
+	logger := controller.LoggerFromContext(ctx).WithName("topology file")
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "topology",
+			Namespace: kuadrantNamespace,
+			Labels:    map[string]string{"kuadrant.io/topology": "true"},
+		},
+		Data: map[string]string{
+			"topology": topology.ToDot(),
+		},
+	}
+
+	configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}
+	unstructuredCM, _ := controller.Destruct(cm)
+
+	targets := topology.Objects().Items(func(object machinery.Object) bool {
+		if object.GetName() == cm.GetName() && object.GetNamespace() == cm.GetNamespace() {
+			return true
+		}
+		return false
+	})
+
+	if len(targets) == 0 {
+		_, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Create(ctx, unstructuredCM, metav1.CreateOptions{})
+		if err != nil {
+			logger.Error(err, "failed to write topology configmap")
+		}
+		return
+	}
+
+	if len(targets) > 1 {
+		logger.Info("multiple topology configmaps found, continuing but unexpected behaviour may occur")
+	}
+	target := targets[0].(controller.Object).(*controller.RuntimeObject)
+	cmTarget := target.Object.(*corev1.ConfigMap)
+
+	if d, found := cmTarget.Data["topology"]; !found || strings.Compare(d, cm.Data["topology"]) != 0 {
```

Comment:
the string comparing is usually very weak and prone to false positives (yielding positive when actually is not). 

Maybe not for now, but a good improvement is to parse the topology (maybe `topology.FromDot(...)` method needed first) and then compare topologies with another method like `parsedTopology.EqualsTo(currentTopology)`.

Reply:
I agree. As the first pass I think this is acceptable. I think the `EqualsTo` function should be part of the Policy-machinery package.